File: MDAnalysis/ramachandran.py
# Benchmark the calculation of Ramachandran phi/psi angles from a PDB file

import time
import MDAnalysis as mda

pdb_filepath = "data/1AKE.pdb"
u = mda.Universe(pdb_filepath)

def ramachandran():
    phi_angles = []
    psi_angles = []
    for res in u.residues:
        try:
            phi = res.phi_selection()
        except:
            pass
        else:
            if not phi is None:
                phi_angles.append(phi.dihedral.value())
        try:
            psi = res.psi_selection()
        except:
            pass
        else:
            if not psi is None:
                psi_angles.append(psi.dihedral.value())
    return phi_angles, psi_angles

start = time.time()
ramachandran()
end = time.time()

print(end - start)

File: Rpdb/ramachandran.R
# Benchmark the calculation of Ramachandran phi/psi angles from a PDB file

library(Rpdb)
library(microbenchmark)

pdb_filepath <- "data/1AKE.pdb"
struc <- read.pdb(pdb_filepath)

ramachandran <- function() {
    is_n <- which(struc$atoms$elename=="N")
    is_ca <- which(struc$atoms$elename=="CA")
    is_c <- which(struc$atoms$elename=="C")
    res_count <- length(is_ca)
    phi_angles <- dihedral(struc, is_c[1:res_count-2], is_n[2:res_count-1], is_ca[2:res_count-1], is_c[2:res_count-1])
    psi_angles <- dihedral(struc, is_n[2:res_count-1], is_ca[2:res_count-1], is_c[2:res_count-1], is_n[3:res_count])
    return(phi_angles, psi_angles)
}

bench <- microbenchmark(ramachandran(), times=1)

cat(bench$time / 10^9, "\n", sep="")
